### ISSUE TITLE
F18 LTD/R 2position Switch Bugged

### ISSUE THREAD

**Issue by Lowlyslows** [DCS-Skunkworks/dcs-bios#134]
Hello,

After much trial and error trying to get the F18 LTD/R switch to turn off when I turn my physical switch off I believe I found a bug. I am using flightpanels DCS-BIOS. When I open the control reference to "toggle" the LTD/R it moves to the ARM position. If I toggle it again it doesn't move. The LST/NFLR moves to on and off when I press the toggle button. If I click on INC the LTD/R switch toggles between Arm and Safe. DEC doesn't do anything. If I click on INC for the LST/NFLR it moves to ON only and if I click DEC it moves to off. It appears that the LTD/R logic is bugged. If I wire my switch to the LST/NFLR and use the same Switch2Pos code but for the LST, the switch will turn on when I move my physical switch to ON and off when I move my physical switch to OFF. LTD/R does not behave this way, it stays on when I turn my physical switch off.

The F18 has auto lasing. So when you do a bomb run the jet calculates the last 15-20 seconds till impact and then starts lasing. Once the bomb impacts the computer kills the power to the magnet and the switch (LTD) moves back to safe. I am trying to replicate that. Not sure if fixing the on/off (arm/safe) to behave correctly will get me to my goal but it’s a start.

thanks!

**Comment by charliefoxtwo** [comment]
Line in question:
https://github.com/DCSFlightpanels/dcs-bios/blob/018b13a440381842cc02a672c3e805f08f719082/Scripts/DCS-BIOS/lib/FA-18C_hornet.lua#L1112

I don't know for certain, but I suspect it should be defined as `defineElectricallyHeldSwitch` similar to the A-10's SAS switches
https://github.com/DCSFlightpanels/dcs-bios/blob/018b13a440381842cc02a672c3e805f08f719082/Scripts/DCS-BIOS/lib/A-10C.lua#L706-L709

@Lowlyslows , have you tried any other electrically-held switches? Would you happen to know all of the electrically-held switches in the aircraft?

**Comment by WarLord211** [comment]
@charliefoxtwo wont work like in the A-10 because the 2 command is missing.

@Lowlyslows wich BIOS version do u use?

**Comment by Lowlyslows** [comment]
@charliefoxtwo hey! Yea I bet you need to have the pod install (station 5). I can get the electromagnet to engage but might just be lucky with code example I stole from the warthog project. If you check that site, you’ll see the magnetic switch I copied. Works well!

**Comment by charliefoxtwo** [comment]
> @charliefoxtwo wont work like in the A-10 because the 2 command is missing. 

@WarLord211 what do you mean by "the 2 command"?

**Comment by WarLord211** [comment]
exactly by A10 the 2. number is the release command.

**Comment by Lowlyslows** [comment]
@WarLord211 I’m confused. Is there something I need to do that’s similar to the A10?

**Comment by charliefoxtwo** [comment]
This is a limitation of the DCS module, so there's not much you can do. I'm trying to find a way to work around it.

**Comment by Lowlyslows** [comment]
@charliefoxtwo what do you mean?

**Comment by charliefoxtwo** [comment]
So as the A-10 is programmed, for switches like the SAS switch there are actually two separate commands, one for on and one for off. For the F-18, it seems there's only one command for toggling the switch. Again, this has to do with the module itself, nothing on the bios side. We're just interfacing with the module.

Are there any other electrically-held switches in the hornet? I know the engine start is a 3 pos electrically held switch but that's it.

**Comment by Lowlyslows** [comment]
@charliefoxtwo I believe the APU switch is. The nice the engines are running the switch turns off. I’ll look for more. Very bummed

**Comment by charliefoxtwo** [comment]
That's very helpful. Can you open up `lib/FA-18C_hornet.lua` and replace line 1112 with the below line and let me know if that solves your problem?
```lua
defineToggleSwitchToggleOnly2("LTD_R_SW", 62, 3002, 441, "Sensor Panel", "LTD/R Switch, ARM/SAFE") 
```

I'm not sure exactly what you should do for the arduino code though. Presumably one of the following?
```cpp
const byte ltdRSwPins[2] = {PIN_0, PIN_1};
DcsBios::SwitchMultiPos ltdRSw("LTD_R_SW", ltdRSwPins, 2);
```
```cpp
DcsBios::Switch2Pos ltdRSw("LTD_R_SW", PIN);
```

**Comment by Lowlyslows** [comment]
@charliefoxtwo is that C:\Users\user\Saved Games\DCS.openbeta\Scripts\DCS-BIOS\lib?

**Comment by Lowlyslows** [comment]
@charliefoxtwo changing that line worked! I need to test the magnet next!!

**Comment by Lowlyslows** [comment]
@charliefoxtwo ok, we are 95%!!! I can now toggle the switch correctly. When I turn the physical switch ON the LTD/R switch turns ON. When I turn the physical switch OFF the LTD/R switch turns OFF. When I turn the physical switch on my electromagnet turns ON and holds the physical switch ON. When I turn the physical OFF the electromagnet turns OFF. When the physical switch is ON (LTD/R switch is ARM) and turn the LTD/R switch (game switch) to SAFE the electromagnet turns OFF and the physical switch turns OFF (pretty cool). 

What still needs to be made to work is when I turn the FLIR to STBY or OFF in the game, the LTD/R switch goes from ARM to SAFE. However, if I have the physical switch turned to ON (LTD/R is ARM) and I turn the FLIR OFF (in game), the switch doesn't de-magnetize and turn OFF. Below is my code:

#define DCSBIOS_IRQ_SERIAL

#include "DcsBios.h"

int Coil = 6;    // Magnetic coil on



DcsBios::Switch2Pos ltdRSw("LTD_R_SW", 2); //LTD ARM/SAFE
DcsBios::Switch2Pos lstNflrSw("LST_NFLR_SW", 3);  //LST ON/OFF
DcsBios::Switch3Pos flirSw("FLIR_SW", 4, 5);  //FLIR ON/STBY/OFF


DcsBios::Switch2Pos hookLever("HOOK_LEVER", 12);  //HOOK up/down

// LTD
void onLtdRSwChange(unsigned int newCoilValue) { //Controls the output of pin 6 and turns on electromagnet
    switch (newCoilValue){
      case 0:
      digitalWrite(Coil, LOW);
      break;
      case 1:
      digitalWrite(Coil, HIGH);
      break;
    }  
}

DcsBios::IntegerBuffer ltdRSwBuffer(0x74c8, 0x4000, 14, onLtdRSwChange);

void setup() {
  DcsBios::setup();
  pinMode (Coil, OUTPUT);
}

void loop() {
  DcsBios::loop();
}

**Comment by Lowlyslows** [comment]
Just tested the logic. I set the LTD/R (game switch to ARM). Doing so turned ON the electromagnet (the physical switch was set to OFF). I did an auto bombing run and after the bomb struck the target (about 3-5 seconds) the LTD/R (game switch) went to SAFE. When it did this it also turned OFF the electromagnet.

**Comment by charliefoxtwo** [comment]
> When the physical switch is ON (LTD/R switch is ARM) and turn the LTD/R switch (game switch) to SAFE the electromagnet turns OFF and the physical switch turns OFF (pretty cool).

> Something about the physical switch being ON is preventing the game logic from turning off the power to the electromagnet.

These two statements seem to conflict with each other 🤔 

> What still needs to be made to work is when I turn the FLIR to STBY or OFF in the game, the LTD/R switch goes from ARM to SAFE. However, if I have the physical switch turned to ON (LTD/R is ARM) and I turn the FLIR OFF (in game), the switch doesn't de-magnetize and turn OFF.

When you do this, what is the value of the LTD/R switch in the bios reference tool?

**Comment by Lowlyslows** [comment]
@charliefoxtwo It's a bit confusing to write, ha!. The "Something about..." has to do with the text right above it. Essentially if I manually turn the LTD/R switch from ARM to SAFE then my electromagnet will de energize. If the game calls to logically turn the game LTD/R switch from ARM to SAFE then the code wont de energize my electromagnet. Again, something about the physical switch being ON is preventing the game logic from turning off the power to the electromagnet.

Not sure how to check the BIOS reference, how do I do this? Is this the Chrome thing?

**Comment by charliefoxtwo** [comment]
Yup, you can check with the chrome tool or with [my tool](https://github.com/charliefoxtwo/Bort), both work (I believe it's under the "Sensors" category).

So by "game logic" you mean like when the switch is deactivated _indirectly_ (not by clicking on _it_, but rather by some other state changing, e.g. master arm off, a/g mode off, flir off, etc)?

**Comment by Lowlyslows** [comment]
@charliefoxtwo yep, that’s exactly what I mean!

**Comment by Lowlyslows** [comment]
Looks like there are several other switches that are electronically held. 

https://forum.dcs.world/applications/core/interface/file/attachment.php?id=172025

**Comment by WarLord211** [comment]
should i check the changed line in?

**Comment by charliefoxtwo** [comment]
@WarLord211 still looking into the other issue that was mentioned. I'll submit a PR once everything is ready :)

**Comment by Lowlyslows** [comment]
@charliefoxtwo @WarLord211: Using the Chrome Reference Tool with game switches only: When I set the LTD/R (game switch) switch to ARM the Output type: integer Current Value is 1 (100%). With it set to SAFE its 0 (0%). When the FLIR (game switch) is set to ON the Current value is 2 (100%), STBY is 1 (50%), and OFF is 0 (0%). When LTD/R (game switch only) is set to ARM and I set the FLIR to STBY the LTD/R current value goes from 1 (100%) to 0 (0%), the FLIR when set to STBY goes from 2 (100%) to 1 (50%).

Using the Chrome Reference Tool with my (physical) switch: When I set the LTD/R physical switch to ARM the current value is 1 (100%) and is being magnetically held at ARM. When I toggle the FLIR (game switch) switch to STBY the FLIR current goes to 1 (50%) but the LTD/R current remains at 1 (100%).

**Comment by Lowlyslows** [comment]
@charliefoxtwo I also confirmed that the APU acts exactly the same way when it should release the electromagnet on my physical switch. Cheers

**Comment by Lowlyslows** [comment]
@charliefoxtwo I tested the A10C for the L SAS switch. I don't know when it logically turns off the switch but I did find something interesting. Everything between the F18 and the A10C functions the same both in and out of the game with all of the above. However, both planes when the generators are turned off de-magnetize the switches (in game). Oddly enough, when I turn off the generators on the A10C my physical switch turns off (electromagnet de-energized). When I turn off the L GEN on the F18, my physical switch remains ON (electromagnet energized).

The other thing to note on the A10C is that when the Emergency Disconnect Lever button is pressed all of the SAS switches turn to off. When I have my physical switch turned on (electromagnet energized) and I hit the emergency disconnect lever key bind, my electromagnet de-energizes. 

Don't know if that test helps at all. 

Merry Christmas


Code I used (same as the F18)

#define DCSBIOS_IRQ_SERIAL
#include "DcsBios.h"


int SaspYawSasL_Coil = 6;    // Magnetic coil


DcsBios::Switch2Pos saspYawSasL("SASP_YAW_SAS_L", 2);

// Yaw SAS Left
void onSaspYawSasLChange(unsigned int newCoilValue) {
    switch (newCoilValue){
      case 0:
      digitalWrite(SaspYawSasL_Coil, LOW);
      break;
      case 1:
      digitalWrite(SaspYawSasL_Coil, HIGH);
      break;
    }  
}

DcsBios::IntegerBuffer saspYawSasLBuffer(0x1108, 0x0400, 10, onSaspYawSasLChange);


void setup() {
  DcsBios::setup();
 
  pinMode (SaspYawSasL_Coil, OUTPUT);
}

void loop() {
  DcsBios::loop();
}

**Comment by charliefoxtwo** [comment]
> The other thing to note on the A10C is that when the Emergency Disconnect Lever button is pressed all of the SAS switches turn to off. When I have my physical switch turned on (electromagnet energized) and I hit the emergency disconnect lever key bind, my electromagnet de-energizes.

This is in the A-10, right? Unfortunately since the A-10 and F-18 are different modules even though you might _think_ the behavior would be the same (especially since they're both by ED), that's not guaranteed to be the case.

Are there any switches in the F-18 that properly de-energize the electromagnet when they are disabled via game logic? My understanding right now is that you haven't found an F-18 switch that will de-energize your electromagnet when the physical switch is on and the switch is turned off via game logic (not by clicking, but by some other action) - is that correct?

**Comment by Lowlyslows** [comment]
@charliefoxtwo id have to run tests on all of them which I am willing to do. Your statement is correct. So far I have tested the LTD and the APU switches. APU is interesting as when the game logic tries to turn off the APU switch (after both engines are running) the switch in game stays on but the LED turns off (physical switch remains ON and magnetized thus keeping the game APU switch ON).

Merry Christmas!

**Comment by charliefoxtwo** [comment]
I know it's a bit of work but it would be really helpful if you were able to. No rush, I'll be busy the next couple days but I should have a bit of free time later in the week to dig deeper and see if we can figure out any way around this issue.

**Comment by Lowlyslows** [comment]
@charliefoxtwo you got it! Glad we can make this better for others as well.

**Comment by Lowlyslows** [comment]
@charliefoxtwo See the link below. I can also copy and paste here as well. I did find something interesting going on. Let me know what you think and if you have any more questions! 

Hope you had a great Christmas!


[Electromagnet Switches.docx](https://github.com/DCSFlightpanels/dcs-bios/files/10304514/Electromagnet.Switches.docx)

**Comment by Lowlyslows** [comment]
@charliefoxtwo something I noticed between the A10C and the F18 lib files. 

A10C: has local defineElectricallyHeldSwitch = BIOS.util.defineElectricallyHeldSwitch which is tied to the SAS switches that I confirmed worked correctly.
F18: Doesn't have this local definition in it, couldn't we add this to the LUA file local definition and replace 
defineToggleSwitchToggleOnly2("LTD_R_SW", 62, 3002, 441, "Sensor Panel", "LTD/R Switch, ARM/SAFE") 

with

defineElectricallyHeldSwitch("LTD_R_SW", 62, 3002, 441, "Sensor Panel", "LTD/R Switch, ARM/SAFE") 


I'm probably way out there and should wait on you, just thinking out loud.

Thanks

**Comment by charliefoxtwo** [comment]
Yeah, I thought that initially here https://github.com/DCSFlightpanels/dcs-bios/issues/134#issuecomment-1363612177

The issue is, that function definition requires two separate argument ids - one for switch up, and one for switch down.

As an example, in the A-10:
```lua
defineElectricallyHeldSwitch("SASP_YAW_SAS_L", 38, 3003, 3004, 185, "SAS Panel", "Yaw SAS Left OFF - ENGAGE") 
```
this has **two** arguments - `3003` and `3004`. 

whereas in the F-18:
```lua
defineToggleSwitch("LTD_R_SW", 62, 3002, 441, "Sensor Panel", "LTD/R Switch, ARM/SAFE") 
```
The only argument here is `3002`.

We control the definition of `defineElectricallyHeldSwitch` so we can (and I've attempted to) modify it or create a copy to support this case. But unfortunately, the switch also behaves differently in DCS because it only has the one argument instead of two, so that doesn't work either.

Other electrically-held switches in the F-18 are defined as `defineToggleSwitchToggleOnly2`, which is why I had you try that to see where that behavior would get us (since it seemed to work for me).

**Comment by Lowlyslows** [comment]
@charliefoxtwo ah ok gotcha. Still learning what the lines of code mean. I ended up trying it and the switch didn't act correctly in the game. When I would move the physical switch to ARM it would move the game switch to ARM but when I released the physical switch back to SAFE the game switch would stay at ARM. If I moved the physical switch back to ARM it would then move the game switch back to SAFE. Also wouldn't energize my electromagnet.

What do you think about the red lines at the beginning of the doc I attached? Interesting that it functions correctly if I "trick" it.

**Comment by charliefoxtwo** [comment]
This is some pretty interesting behavior indeed.

> Chances are, all the other switches below will function the same way (ON in game first and the ON in real life to be logically turned OFF by the game)

Were there any that you tested that _didn't_ behave this way?

**Comment by Lowlyslows** [comment]
@charliefoxtwo i can test them as well but they all showed the same behavior before I figured this out so my assumption is yes. Let me know if you need me to confirm

**Comment by Lowlyslows** [comment]
@charliefoxtwo tested them and they all show the same behavior. I have to turn the switch ON first in the game, then turn my physical switch ON, and then the game will logically turn OFF my physical switch.

**Comment by charliefoxtwo** [comment]
Got it. It sounds like (just rambling here, this doesn't mean anything) when the switch is toggled on physically, some state value in dcs that gets changed when the switch is clicked in-game isn't being changed - but for some reason, this isn't an issue with the A-10 switches.

I still have no idea _why_ this would be the case, but more information is always helpful. Hopefully we can get to the bottom of this.

**Comment by Lowlyslows** [comment]
@charliefoxtwo yep. I was thinking on the drive into work that this has something to do with the game switch being commanded ON vs the command to turn ON from DCS-BIOS. Let me know what you would like me to test next!

**Comment by Lowlyslows** [comment]
@charliefoxtwo more testing: the "Trick" behavior is the same when I use a physical switch to control the FLIR. Meaning, I can set the physical FLIR switch to ON/STBY/OFF without doing it prior with the game switch. Then I do the "trick" with the LTD switch. After that, if I move the physical FLIR switch from ON to STBY it will move the game FLIR switch to STBY and then de-energize the physical electromagnet and move both the LTD game switch and physical switch to SAFE.

**Comment by Lowlyslows** [comment]
@charliefoxtwo hey! Thinking out loud here. Is there a way to have the code check and see the status of things that would affect the electromagnet from powering on? For example, if the FLIR is in STBY or OFF then the LTD/R electromagnet won't energize.

**Comment by charliefoxtwo** [comment]
Yeah, but we try not to build in these dependencies for each individual switch. Sometimes game logic can also disable switches without a corresponding switch (e.g. maybe the switch automatically disables after munition impact, regardless of the flir switch - not saying it's the case for this switch, but could be for other switches).

**Comment by Lowlyslows** [comment]
> Yeah, but we try not to build in these dependencies for each individual switch. Sometimes game logic can also disable switches without a corresponding switch (e.g. maybe the switch automatically disables after munition impact, regardless of the flir switch - not saying it's the case for this switch, but could be for other switches).

Makes sense. I have spent the past week trying to figure out why the switch won’t return logically like the A10. I have been trying to stay out of the LUA files as last time I broke DCS and had to reinstall it. I was thinking of comparing the game module LUAs (18 and A10) to see what’s different but haven’t gotten around to it yet. 

On a positive note, it’s been fun to code and see stuff work!!🤣

**Comment by DeadMeatDCS** [comment]
Agree with @charliefoxtwo we shouldn't add switch logic in DCS-BIOS that the already game controls.

@Lowlyslows Mag switches in the F18 are simply implemented in DCS in a different way than for the A10 which is why the old approach to controlling them doesn't work. It's a known problem (well known to some I guess).

If you use DCS-BIOS to send the command to toggle the switch (LTDR, APU, any) you are forcing the switch into the ON position. This is similar to binding it to a keystroke and holding in the key and never releasing it. This can only be overridden by clicking the switch with the mouse in game - or sending the DCS-BIOS command again.

This is why your mag switch is not automatically turning off when you expect it to (e.g.: turning off the battery should turn off the APU, turning off FLIR should turn off LTDR... etc.) It doesn't have anything to do with FLIR or anything. It is simply because you're effectively holding the LTDR switch in by way of DCS-BIOS.

If we stick with the keyboard approach, you'll notice that you can indeed force a switch on (e.g. LTDR), but as soon as you release the key it flips back off IF it should stay off by game logic (e.g. FLIR is off) - or it stays on IF it is allowed.

**This is the key to the known, current workaround: use a Arduino Pro Micro (or Leonardo) to emulate a HID keyboard (or joystick) to send a keypress (and key release!!!) that you've bound to the switch in game.** You can run DCS-BIOS in the same sketch to monitor the switch position to control your physical magnet.
With a bit of logic added you can control and stay in sync with the sim..

**Comment by DeadMeatDCS** [comment]
To clarify: the LTDR switch is not bugged in DCS or in DCS-BIOS as such. You're seeing a general problem with mag switches in the Hornet compared to the Warthog. Same applies for the other 2-position mag switches.

3-position mags are cursed in their own way to boot - but all can be handled with the HID keyboard/joystick workaround until someone cooks up a better solution.

I'd recommend to close the issue @WarLord211 and @Lowlyslows or maybe better yet rephrase into a general issue around how to deal with mag switches, not just the LTDR. We need a LUA hero here :)

**Comment by charliefoxtwo** [comment]
Yup, I'm aware of the issue, but the point we've gotten it to (everything works except for when the **physical** switch is on _and_ game logic toggles it off) suggests to me that there is _something_ we can do to get it working, I'm just not sure what yet.

**Comment by kbastronomics** [comment]
@DeadMeatDCS,  I would agree but not.   with a closed issue you have little ammunition to take to DCS as you're saying you have a work around.   we need this to function correctly.  with the Pro-Micro since it supports native HID (which is good) but until DCS-BIOS is fixed we can't use RS-485 which is core to the design of OpenHornet..   so it's a catch 22 for the official OH code.   for others rolling their own its an option for sure.

**Comment by kbastronomics** [comment]
@charliefoxtwo 
FYI,  when I watch both thru status LEDS and on the scope I do see the signal come across to disable the switch magnet via the onApuControlSwChange and onEngineCrankSwChange.  I've only played with the APU board at this point as it's the one I've built so far.

**Comment by Lowlyslows** [comment]
@charliefoxtwo pro micro gets here tomorrow. With the “trick” behavior previously discussed I am sure this will work. Glad others have figured it out already, wish I had ran into open hornet 2 weeks ago🤣

**Comment by DeadMeatDCS** [comment]
Agree it is not really figured out. Though I maintain that the issue is more general to the F18 and does not only apply to the LTDR switch as the issue suggests. But whatever :)

Here's some input. There was a list of mag switches shared above. Disregarding what may or may not be mag switches the real jet, the following 13 toggles + 1 knob pull are actually implemented as mag switches in DCS, and may call for slight variations in how they are "fixed" due to how they are defined. Test them yourself if you'd like.

_ToggleSwitch_
- LTD/R
- Bleed air aug pull (not a toggle switch as such: pull the top knob)

_ToggleSwitchToggleOnly2_
- APU
- Fuel dump (special case: this is actually meant to be allowed to be held on manually without mag hold, to allow fuel dump at any time, e.g. below BINGO)
- Hook bypass
- Launch bar
- Pitot heat

_RockerSwitch_
- Ground power 1-4
- Engine crank
- Canopy (though only mag held in up position)

_MissionComputerSwitch_
- MC switch (on hyd isol panel)

It could be argued that the gear switch should be included since there's the downlock solenoid keeping it in down postion with WoW... I'll leave that up to someone else to decide :)

**Comment by charliefoxtwo** [comment]
The canopy switch may not be using the correct one either if it's only held in one position, but I'll have to check that separately.

The landing gear wow thing is basically the reverse - in normal cases the electromagnet prevents the switch from toggling on its own, but still allows the pilot to toggle the switch. This is reversed in that the switch won't toggle on its own, and it prevents the pilot from toggling it. More investigation required for that one as well.

**Comment by Lowlyslows** [comment]
@charliefoxtwo Got the pro micro working. Issue is: I can only get the LTD switch to move to ARM. When I move the physical  switch back from ARM to Safe, it stays at ARM.
Also, I noticed something. When I use the physical switch to move the LTD game switch to ARM and then move the FLIR to STBY, the LTD electromagnet doesn't de-energize. If I use the keyboard and bind the ARM position to the S key, press the S key which moves the LTD game switch to ARM and release the S key, moving the FLIR switch to STBY will de-energize the LTD electromagnet. However, if I hold the S key down and then move the FLIR switch to STBY, the electromagnet stays energized. This leads me to believe that my physical switch is HOLDING the bind active and not allowing the logic to de-energize the electromagnet.

**Comment by Lowlyslows** [comment]
added this to the Mods\aircraft\FA-18C\Input\FA-18C\joystick\default.lua just below line 840

Now the switch turns on and off with a single command on/off toggle switch (single pole). Now I need to figure out how to make the command to be momentary and not hold the bind.

--added 1 line below 03JAN23 
{    down = tgp_commands.LtdrArm,                up = tgp_commands.LtdrArm,                        cockpit_device_id = devices.TGP_INTERFACE,        value_down =  1.0,    value_up = 0.0,    name = _('LTD/R Switch (special) - ON/OFF'),                    category = {_('Special For Joystick'), _('Right Console'), _('Sensor Panel')}},

**Comment by kbastronomics** [comment]
@charliefoxtwo,   I can reproduce this problem now without a physical switch.
launch DCS and the reference utility.  go to live data.   do a cold start, Battery, skip fire test for now, 
Select F/A-18 and APU Panel.
Use the Set State control in the reference utility and turn on the APU  Set State 1,(do not do thru the cockpit itself.)
use the Set State again to Crank the Right Engine. Position (2)
Bring RIght engine to Idle
Once at idle the state of the Crank Switch should change to the Center Position (1) which would release the mag switch.  
It does not

Next, I reset things to a cold/dark hornet and this time 
I started the APU and Cranked the Right engine using the on-screen toggles.
I then set state as above,
The toggles returned to normal and worked as they should even thought they were set state the same but only after using the cockpit toggles.

I'm not sure where this actually point the issue to be at but it is not the actual switch being used that causes it.

**Comment by charliefoxtwo** [comment]
I wonder if there's something to the theory that bios is somehow holding it in place. I'll have to dig deeper into bios to figure out why that would be the case, as I wasn't one of the original low-level authors.

**Comment by DeadMeatDCS** [comment]
@charliefoxtwo DCS-BIOS does indeed seem to try to set the position. I cannot find the quote but I believe Ian (original author) said as much. It is not sending keystrokes (and releases) but more like directly interacts with the .... uh control lua stuff. It'd be nice to confirm where and how this is happing in the source code I agree.

For the majority of switches this is fine and it doesn't matter. Actuating physical switches always results in the same result in DCS. Only at start-up can there be a discrepancy between physical and sim since dcs-bios only detects changes. I saw a solution to this however with some code that detected a new "module" session start and then triggered changes where there were position discrepancies.. Somewhere on ED forums.

**For our dear mag switches, however there is the problem.** If you use a physical switch as a trigger for change (or live reference, same deal right @kbastronomics ) and then read position in DCS (output via DCS-BIOS) to control the solenoid you end up with the issue: a self fulfilling condition. Solenoid engaged because the switch is held in. This is also true for other modules with mag switches like the Harrier and the Tomcat. Only the A10 mag switches are different - but how??

A related issue is when trying to set a position for a switch that requires some other condition (toggle switches with switch guards, parking brake that must be rotated first, landing gear switch with downlock, radar EMER position rotary that needs to be lifted first). Physically you set a position and dcs-bios sends a command but the switch doesn't oblige in DCS... Solutions here require either trigging the conditions by separate physical switches and sending appropriate commands or trigging a macro like sequence that does the work (e.g. lift the guard, then toggle the switch when the guard is detected open/a certain time has elapsed; all triggered by just the physical toggle switch)

**Should these issues be solved in dcs-bios?** A matter of taste I guess - but in it's purest form dcs-bios can detect all of these positions and set them. Overriding game logic or creating macro like sequences is probably better left to the user to handle in either code or physically with switches than forcing a specific solution.

**Comment by DeadMeatDCS** [comment]
@kbastronomics  I know you're asking @charliefoxtwo but here you go:
the 300x commands and the switch id 377 is documented in lua files within aircraft sub directories
DCS directory\Mods\aircraft\FA-18C\Cockpit\Scripts

**command_defs.lua** contains the commands (basically a counter for each function). You'll see APU is 3001 in the same section.
clickabledata.lua has the definition of the switch type and function, referencing the 300x commands and an ID, 377 for crank

**Comment by DeadMeatDCS** [comment]
Lights and instrument related stuff is found in the other sub folders.. e.g. lamps.lua under \MainPanel

**Comment by kbastronomics** [comment]
This is just a test of a theory and since my APU panel is still not ready
can someone try this in loop()  just add a delay

void loop() {
  DcsBios::loop(); 
  delay(100L);
}

It may or may not do anything.   but I'm curious

**Comment by kbastronomics** [comment]
Oh and BTW:   doing a cold start and using the Control Reference I can make it work if I do this
do normal cold start,  bat sw one,  can skip fire test for this
start the APU 
Start Right Engine
Engine Crank Switch, LEFT/OFF/RIGHT set state 2 then immediately 
Engine Crank Switch, LEFT/OFF/RIGHT set state 1 
bring engine to idle when ready
you will now see the Engine Crank Switch return to the OFF position as it should once the engine is up and running

Start Left Engine
Engine Crank Switch, LEFT/OFF/RIGHT set state 0 then immediately 
Engine Crank Switch, LEFT/OFF/RIGHT set state 1 
bring engine to idle when ready
you will now see the Engine Crank Switch return to the OFF position as it should once the engine is up and running

so when the Switch is held in the 0 or 2 position it will not return to the off which is consistent with the issue.

**Comment by charliefoxtwo** [comment]
> **For our dear mag switches, however there is the problem.** If you use a physical switch as a trigger for change (or live reference, same deal right @kbastronomics ) and then read position in DCS (output via DCS-BIOS) to control the solenoid you end up with the issue: a self fulfilling condition. Solenoid engaged because the switch is held in.

Well what's interesting though is that if you click the switch in-game everything will work as expected (i.e. the electromagnet will disengage - correct me if I'm wrong @Lowlyslows). This issue seems to only occur when something _other than clicking the exact switch_ disables the switch.

I'll have to dive deeper. Perhaps there's some deeper logic that will only update a value when the clickable has actually been clicked, and doesn't care if something else is changing it. But still, @Lowlyslows noticed that if the physical switch is off, then the electromagnet will engage and disengage properly. It only doesn't work when the physical switch is on _and_ the electromagnet should be disabled by something other than directly clicking on the switch.

**Comment by Lowlyslows** [comment]
@charliefoxtwo thats correct. I have concluded that when the physical switch is in the ARM position, it is constantly pressed. Same as if you held the S key down you would get sssssssssssssss… and so on.

**Comment by charliefoxtwo** [comment]
> I have concluded that when the physical switch is in the ARM position, it is constantly pressed. Same as if you held the S key down you would get sssssssssssssss… and so on.

But if this were the case, then I would also expect just clicking on the switch in game not to work either - but you said it does work, right? Clicking on the LTD/R switch in game disables the electromagnet, right?

@jdahlblom, I noticed you did the original F-18 module. Would you happen to have any input here?

**Comment by DeadMeatDCS** [comment]
Andreww aka Bullant created the module initially:
https://forum.dcs.world/topic/179939-dcs-bios-fa-18c-library-get-it-here

It was since merged into FP branch here.

**Comment by kbastronomics** [comment]
@charliefoxtwo yes that is what I'm seeing in the control reference as well.
When the switch is interacted with via DCS-BIOS (arduino code can be ruled out since this also happens in the control reference which would not use arduino code.) it will not be released (ie mag switch is not turned off)

if we use either Keyboard interaction or thru the cockpit switches it works as it should.

How do you guy debug the DCS-BIOS code?  I'd like to look deeper if I could

**Comment by charliefoxtwo** [comment]
> How do you guy debug the DCS-BIOS code? I'd like to look deeper if I could

I generally use `BIOS.log()` to write out whatever I want to see. It works alright-ish, but it's not the best developer experience 🤷 

https://github.com/DCSFlightpanels/dcs-bios/blob/3110bfd23200df7a8b62dcc71a6c9a57712d6411/Scripts/DCS-BIOS/BIOS.lua#L8-L13

**Comment by Lowlyslows** [comment]
> > I have concluded that when the physical switch is in the ARM position, it is constantly pressed. Same as if you held the S key down you would get sssssssssssssss… and so on.
> 
> But if this were the case, then I would also expect just clicking on the switch in game not to work either - but you said it does work, right? Clicking on the LTD/R switch in game disables the electromagnet, right?
> 
> @jdahlblom, I noticed you did the original F-18 module. Would you happen to have any input here?

It depends on what I am doing to activate the switch. If I am going through DCS BIOS then yes but now I am going to a Pro Micro and things are different. Figured you wouldn't care about the pro micro. So yes, click on the switch in the game while switch control is done through DCS BIOS will disable the electromagnet.

**Comment by Lowlyslows** [comment]
Out of curiosity, how does the LTD/R switch know to get set to SAFE after an auto bombing run?

**Comment by DeadMeatDCS** [comment]
@Lowlyslows The actual release parameters are handled somewhere in the DCS module code and is not directly visible to us through clickables or instrument outputs.

Same goes for parameters for when to allow mag hold-in. It's coded somewhere that you need power on the right bus, you need to be in A/G mode, have a pod fitted etc... We typically can't read the contributing conditions directly and we (so far) cannot find/see the flag/variable determining hold-in/release.

**Comment by Lowlyslows** [comment]
@DeadMeatDCS thanks for the explanation! Thinking "work around", do you think that if the signal to move the LTD switch in the game from the physical switch was a momentary ON signal and then was coded to turn the signal off would work? Kinda like pressing and releasing a keyboard key. This would, IMO, allow two things to happen. 1) The "button" would not be constantly pressed and 2) allow the logic in the game to function.

I ask because I figured out how to make this action happen with an LED. When the switch turns on the LED flashes which in turn could be the ON/ARM command for the LTD switch. I just don't know how to implement the code into either the HID controller (Pro Micro) or through DCS-BIOS and my Nano. 

Just a thought

**Comment by kbastronomics** [comment]
Question,   which function is best to see incoming messages from DCS?
I'm thinking this one
function BIOS.protocol.processInputLine(line)

Or maybe in the listeners
function BIOS.protocol_io.UDPListener:step()
function BIOS.protocol_io.TCPServer:step()

but if there's a better place I'm game for looking at it.

I'm not seeing any message come into the Rocker switch used for the Engine Crank switch but want to see if a message comes in just not what's expected.

**Comment by kbastronomics** [comment]
Has there been any progress on the root cause of why the mag switches are not working correctly?  I'vr continued to look but wondering if the project team has identified the issue or not?

**Comment by charliefoxtwo** [comment]
I just haven't had the time to investigate this to the depth required lately. Hoping somebody else from the team cna take over. We have all the information, but it just doesn't make sense to me at the moment and I've been too busy with other projects to be able to give this one the time needed unfortunately.

**Comment by kbastronomics** [comment]
I'm trying to figure out the interface to DCS itself for receiving events from DCS.   everywhere I try to trace the program doesn't seem to be where that happens.   

That was why my question above.   I"m happy to work on this and if I find the issue and it is DCS-BIOS that has the issue I'll submit a PR for it.   

I just need to underrstand how DCS-BIOS receives messages from DCS so I can insert some debug code to see what's happening

Can you or anyone share that bit of information?

**Comment by charliefoxtwo** [comment]
I suspect it's in Protocol.lua or ProtocolIO.lua, but I honestly don't know for certain how it gets the data.

**Comment by charliefoxtwo** [comment]
I suspect that change will just get us to where we were after I suggested using a different method: it's a change that needs to be made and it will get us _close_, but I don't _think_ it will fix the game logic issue? Eager to see though.

**Comment by kbastronomics** [comment]
This fixes one specific issue,  not the root cause of the mag switch issue.

One thing to note and others can correct me if I'm wrong but some of the code being tested is not pure DCS-BIOS code using button function but sending direct commands etc to get around issue.

We should be sure to us pure DCS-BIOS code for all testing

[apu_test_code.zip](https://github.com/DCSFlightpanels/dcs-bios/files/10571430/apu_test_code.zip)

Attached is what I use to test the APU panel.   
with the exception of using the sendDcsBiosMessage calls to trick DCS-BIOS into returning the switch to it off position is pure code.  it should be working without the sendDcsBiosMessage needed

**Comment by DeadMeatDCS** [comment]
Haven't tested yet but @kbastronomics looks like you're using the APU ready light to control the solenoid. We don't have something similar for the other 2-position mag switches so the approach doesn't solve it for all cases I'd think..
More testing needed I think

**Comment by kbastronomics** [comment]
@DeadMeatDCS , Correct on the APU ready light,  that was me doing some testing

**Comment by snakeoneone** [comment]
Members
[ 4](https://forum.dcs.world/profile/145595-bessta/content/)
[Posted 16 hours ago](https://forum.dcs.world/topic/334633-dcs-bios-fa-18-bug/?do=findComment&comment=5299957)
DcsBios::Switch2Pos ltdRArm("LTD_R_ARM", PIN);

 

No matter wich switch i map using that line of code i always have it messing my Flir switch. It cycles the flir instead o arming or rendering LTD safe.


it sometimes acts uppon the lst/nflir switch.

**Comment by jdahlblom** [comment]
Where are we with this issue? Solved, partly solved, no solution available?

**Comment by kbastronomics** [comment]
DCS-BIOS is not constructed in a way that will allow this to work properly.   we've tried and dcs-bios is not sending and/or receiving commands that allow the MAG switches to work as they should.

**Comment by jdahlblom** [comment]
Recap : 
First post indicated that the ```LTD_R_SW``` didn't work. Works as expected using ```BIOSBuddy``` *as in* you can ```INC```/```DEC``` and ```TOGGLE```.

There was also a wish to have this switch return from ```ARM``` to ```SAFE``` after the target computer turns off lasing. This isn't possible.

Is this a correct recap?

**Comment by snakeoneone** [comment]
The magnet stuff i don't know because i don't plan on giving it a try.After uninstalling and updating dcs-bios i was able to get lst//nflr to work without cycling my flir 3 position switch

Yahoo Mail: Pesquise, organize, aproveite 
 
  Em quinta, 7 mar., 2024 às 15:30, Jerker ***@***.***> escreveu:   


Recap :
First post indicated that the LTD_R_SW didn't work. Works as expected using BIOSBuddy as in you can INC/DEC and TOGGLE.

There was also a wish to have this switch return from ARM to SAFE after the target computer turns off lasing. This isn't possible.

Is this a correct recap?

**Comment by jdahlblom** [comment]
Closing this as the control now works as DCS-BIOS controls are supposed to work.

**Comment by kbastronomics** [comment]
Just a comment,  MAG Switches do not work which was part of the original issue

"The F18 has auto lasing. So when you do a bomb run the jet calculates the last 15-20 seconds till impact and then starts lasing. Once the bomb impacts the computer kills the power to the magnet and the switch (LTD) moves back to safe. I am trying to replicate that. Not sure if fixing the on/off (arm/safe) to behave correctly will get me to my goal but it’s a start."

So this is not completed as that still does not work as it should without coding work arounds.   which is basically bypassing DCS-BIOS and using HID controls for the 7 mag switches in the F/A-18 by making them either button presses via a joystick control or a keypress event.

if this is being close because DCS-BIOS does not plan on fixing that issue (which i understand is a core item in the way that DCS-BIOS works) thats fine but that should be spelled out in the closure.

**Comment by charliefoxtwo** [comment]
The way mag switches work in dcs-bios is you have to trigger them on (say to the 1 position) and then trigger them off (say to the 0 position) as if they're a push button. If you only trigger them on, then the game basically acts like your finger is still on the switch and even though it's magnetically released it won't go back to 0.

We tried to find a way with a single bios command to send something equivalent to the trigger on and then release, but that didn't seem possible.

Unfortunately, there's just no way for us to get all the information here since we can't fetch the magnet's status and we're trying to make do with just the switch's position.

Tl;dr: send 1, then send 0. This might need to be done with a delay - I don't know. That should leave the switch on and locked. To manually turn it off, send 0 again.

**Comment by kbastronomics** [comment]
That was the first thing we tried. It doesn't work, at least on the hornet going through the Areno library and doing that the mag switch will never disengage doing the same thing in the CNTLREF or BORT works fine, but not with the real switch and magnet.

**Comment by kbastronomics** [comment]
Tried that as well we started with a 10 ms delay then 50 then a 250 then all the way up to a one second delay no change.

**Comment by charliefoxtwo** [comment]
It's suspicious to me that it works in ctrlref/bort but not in the arduino library since they do the exact same thing (presuming you're using the arduino code that calls into the same interface as the bort control you tested - i.e. maybe standard inc/dec worked but set_state didn't or vise-versa)

**Comment by kbastronomics** [comment]
what we found was every control in dcs-bios Arduino where the switch is read directly would not work.
if we caused a keypress thru hid it worked, also thru a joystick button.   
but when the actual switch was read it would never release.
in all cases we used the output from the switch to turn on/off the magnet.

**Comment by BojoteX** [comment]
I solved it by replacing:

`FA_18C_hornet:defineToggleSwitch("LTD_R_SW", 62, 3002, 441, "Sensor Panel", "LTD/R Switch, ARM/SAFE")`

with:

`FA_18C_hornet:defineToggleSwitchToggleOnly("LTD_R_SW", 62, 3002, 441, "Sensor Panel", "LTD/R Switch, ARM/SAFE")`

Notice the "ToggleOnly" part... this was after some trial an error, noticed the pitot switch had this setting, decided to try it and got lucky. Just to clarify, what this solves is the issue of the LTD/R switch not going back to SAFE when cycling it after an auto lase/bombing run. As some of you have noticed, you need to cycle it 3 times in order to get back to "ARM" as SAFE will not engage on the first try

**Comment by charliefoxtwo** [comment]
Reopening as this isn't fixed in dcs-bios. I'm also hesitant to switch to the toggle-only interface as we then lose the inc/dec and set state commands (plus, I don't see any reason toggle would only work in toggle only, since the toggle commands are the same in both control definitions...)

**Comment by kbastronomics** [comment]
LTDR is a Magnetically held switch, the code above for a mag switch will not work (the A10 code does not work for the hornet, we do not have the same electrically held switch definition in DCS-BIOS.   

There's a issue in DCS-BIOS and the Hornet that causes all mag switches to require custom coding to work.   just relying on the callback function alone will not work for the hornet 

generally, you'll see the mag switch engage but never disengage but I believe there's been other situations.
a change to the order lying code in DCS-BIOS may cause those using non magnetically held switch's to no longer function so we have to be careful and make sure any change works in all conditions.

**Comment by charliefoxtwo** [comment]
Closing as fixed in #1627
